src/merkle_tree/mod.rs

```diff
@@ -1,253 +1,403 @@
+    pub fn new<L: ToBytes>(
+        leaf_hash_param: &LeafParam<P>,
+        two_to_one_hash_param: &TwoToOneParam<P>,
+        leaves: &[L],
+    ) -> Result<Self, crate::Error> {
+        let leaf_nodes_size = leaves.len(); // size of the leaf layer
+        assert!(
+            leaf_nodes_size.is_power_of_two(),
+            "`leaves.len() should be power of two"
+        );
+        let non_leaf_nodes_size = leaf_nodes_size - 1;
+
+        let tree_height = tree_height(non_leaf_nodes_size + leaf_nodes_size);
+
+        let hash_of_empty: TwoToOneDigest<P> = P::TwoToOneHash::evaluate_both(
+            two_to_one_hash_param,
+            &vec![0u8; P::TwoToOneHash::LEFT_INPUT_SIZE_BITS / 8],
+            &vec![0u8; P::TwoToOneHash::RIGHT_INPUT_SIZE_BITS / 8],
+        )?;
+
+        // initialize the merkle tree as array of nodes in level order
+        let mut non_leaf_nodes: Vec<TwoToOneDigest<P>> = (0..non_leaf_nodes_size)
+            .map(|_| hash_of_empty.clone())
+            .collect();
+        let mut leaf_nodes: Vec<LeafDigest<P>> = Vec::with_capacity(leaf_nodes_size);
+
+        // Compute the starting indices for each non-leaf level of the tree
         let mut index = 0;
-        let mut level_indices = Vec::with_capacity(tree_height);
-        for _ in 0..tree_height {
+        let mut level_indices = Vec::with_capacity(tree_height - 1);
+        for _ in 0..(tree_height - 1) {
             level_indices.push(index);
             index = left_child(index);
         }
 
-        // Compute and store the hash values for each leaf.
-        let last_level_index = level_indices.pop().unwrap_or(0);
-        let mut buffer = [0u8; 128];
-        for (i, leaf) in leaves.iter().enumerate() {
-            tree[last_level_index + i] = hash_leaf::<P::H, _>(&parameters, leaf, &mut buffer)?;
+        // compute and store hash values for each leaf
+        for leaf in leaves.iter() {
+            leaf_nodes.push(P::LeafHash::evaluate(
+                leaf_hash_param,
+                &ark_ff::to_bytes!(leaf)?,
+            )?)
+        }
+
+        // compute the hash values for the non-leaf bottom layer
+        {
+            let mut left_bytes;
+            let mut right_bytes;
+            let start_index = level_indices.pop().unwrap();
+            let upper_bound = left_child(start_index);
+            for current_index in start_index..upper_bound {
+                let left_leaf_index = left_child(current_index) - upper_bound;
+                let right_leaf_index = right_child(current_index) - upper_bound;
+                // compute hash
+                left_bytes = ark_ff::to_bytes!(&leaf_nodes[left_leaf_index])?;
+                right_bytes = ark_ff::to_bytes!(&leaf_nodes[right_leaf_index])?;
+                non_leaf_nodes[current_index] = P::TwoToOneHash::evaluate_both(
+                    &two_to_one_hash_param,
+                    &left_bytes,
+                    &right_bytes,
+                )?
+            }
```

Comment:
This logic could be merged with the one below since they are pretty similar

Reply:
I also tried to merge those two when I was writing this. However, there are some difficulties: 
- `leaf_nodes` and `non_leaf_nodes` may have elements of different types if user supplies different CRH for leaf layer and inner node layers. Thus, it is tricky to combine those two lists. 
- `left_leaf_index` and `left_index` use different logics, and if we combine those two, we might need to check if we are at the bottom non-leaf level at runtime for every node.
- We can also define a helper function with `#[inline]` to avoid duplicate code, but the duplicate here is relatively short, so I'd prefer not do that (since adding helper function with many arguments also make it hard to read). 

There might be better ideas. Feel free to add further suggestions and I'm willing to implement them if that makes sense. Thanks!

Reply:
I may take a stab at it in the near future. It looks like it depends on what's higher on your priority list and what it means to have cleaner code. Either way, I'm fine with it

Reply:
Sound good. Thanks for the help!